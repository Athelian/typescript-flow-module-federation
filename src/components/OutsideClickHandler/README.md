Outside Click Handler example

```js
<OutsideClickHandler onOutsideClick={() => console.warn('click outside')}>
  Click outside of this element and see on the console log.
</OutsideClickHandler>
```

Ignore element example

```js
<OutsideClickHandler
  ignoreElements={[document.body]}
  onOutsideClick={() => console.warn('it should not trigger')}
>
  This will not fire onOutsideClick() because we ignore the document.body element.
</OutsideClickHandler>
```
