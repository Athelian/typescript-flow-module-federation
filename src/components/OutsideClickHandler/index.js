// @flow
import * as React from 'react';

type Props = {
  /**
   * List of element we want to ignore.
   * `onOutsideClick()` won't call if click on that element.
   *
   * @type {Array<Node>}
   */
  ignoreElements?: Array<Node>,
  /**
   * Render prop
   *
   * @type {React.Node}
   */
  children: React.Node,
  /**
   * Function will be called when click outside of element
   *
   * @type {Function}
   */
  onOutsideClick: Function,
};

export default class OutsideClickHandler extends React.Component<Props> {
  static defaultProps = { className: '', ignoreElements: [] };

  componentDidMount() {
    document.addEventListener('mousedown', this.onOutsideClick);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.onOutsideClick);
  }

  onOutsideClick = (evt: MouseEvent) => {
    const { ignoreElements, onOutsideClick } = this.props;
    const isOutsideTarget =
      this.wrapperRef && evt.target instanceof Node && !this.wrapperRef.contains(evt.target);
    const isIgnore =
      ignoreElements && ignoreElements.length
        ? ignoreElements.find(
            item =>
              item &&
              ((evt.target instanceof Node && item.isSameNode(evt.target)) ||
                (evt.target instanceof Node && item.contains(evt.target)))
          )
        : false;
    if (isOutsideTarget && !isIgnore) {
      onOutsideClick(evt);
    }
  };

  setChildNodeRef = (ref: HTMLDivElement | null) => {
    if (ref) this.wrapperRef = ref;
  };

  wrapperRef: HTMLDivElement;

  render() {
    const { children } = this.props;
    return <div ref={this.setChildNodeRef}>{children}</div>;
  }
}
