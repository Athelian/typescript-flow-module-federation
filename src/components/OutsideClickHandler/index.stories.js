import React from 'react';
/* eslint-disable import/no-extraneous-dependencies */
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import OutsideClickHandler from './index';

storiesOf('OutsideClickHandler', module)
  .add('with handler', () => (
    <OutsideClickHandler onOutsideClick={action('onOutsideClick')}>
      Click outside of this element then look at action logger.
    </OutsideClickHandler>
  ))
  .add('with ignore elements', () => (
    <div>
      <h3> This will be ignore </h3>
      <OutsideClickHandler
        ignoreElements={[document.getElementById('root')]}
        onOutsideClick={action('it should ignore on #root element')}
      >
        This will not fire onOutsideClick() because we ignore the `#root` element.
      </OutsideClickHandler>
    </div>
  ));
