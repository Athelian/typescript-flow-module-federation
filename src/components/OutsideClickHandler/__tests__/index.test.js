import React from 'react';
import { render } from '@testing-library/react';
import OutsideClickHandler from '../index';

describe('<OutsideClickHandler />', () => {
  it('should render without crash', () => {
    const { container } = render(
      <OutsideClickHandler onOutsideClick={() => {}}>
        Click outside of this element.
      </OutsideClickHandler>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
