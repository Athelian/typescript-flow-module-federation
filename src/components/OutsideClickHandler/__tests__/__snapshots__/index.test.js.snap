// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OutsideClickHandler /> should render without crash 1`] = `
<div
  class=""
>
  Click outside of this element.
</div>
`;
